Targets/STM32F4/STM32F4_CAN.cpp

```diff
@@ -0,0 +1,1597 @@
+
+    int32_t channel = self->Index;
+
+    _lowerBoundFilters = (uint32_t*)memoryProvider->Allocate(memoryProvider, length * sizeof(uint32_t));
+    _upperBoundFilters = (uint32_t*)memoryProvider->Allocate(memoryProvider, length * sizeof(uint32_t));
+
+    if (!_lowerBoundFilters || !_upperBoundFilters) {
+        memoryProvider->Free(memoryProvider, _lowerBoundFilters);
+        memoryProvider->Free(memoryProvider, _upperBoundFilters);
+
+        return  TinyCLR_Result::OutOfMemory;
+    }
+
+    memcpy(_lowerBoundFilters, lowerBounds32, length * sizeof(uint32_t));
+    memcpy(_upperBoundFilters, upperBounds32, length * sizeof(uint32_t));
+
+    bool success = InsertionSort2CheckOverlap(_lowerBoundFilters, _upperBoundFilters, length);
+
+    if (!success) {
+        memoryProvider->Free(memoryProvider, _lowerBoundFilters);
+        memoryProvider->Free(memoryProvider, _upperBoundFilters);
+
+        return TinyCLR_Result::ArgumentInvalid;
+    }
+
+    {
+        DISABLE_INTERRUPTS_SCOPED(irq);
+
+        canData[channel].groupFiltersSize = length;
+        canData[channel].lowerBoundFilters = _lowerBoundFilters;
+        canData[channel].upperBoundFilters = _upperBoundFilters;
+    }
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_DiscardIncomingMessages(const TinyCLR_Can_Provider* self) {
+    int32_t channel = self->Index;
+
+    canController[channel].can_rx_count = 0;
+    canController[channel].can_rx_in = 0;
+    canController[channel].can_rx_out = 0;
+
+    return TinyCLR_Result::Success;
+}
+
+TinyCLR_Result STM32F4_Can_TransmissionAllowed(const TinyCLR_Can_Provider* self, bool &allow) {
+    int32_t channel = self->Index;
+
+    CAN_TypeDef* CANx = ((channel == 0) ? CAN1 : CAN2);
+
+    allow = false;
+
+    if ((CANx->TSR&CAN_TSR_TME0) == CAN_TSR_TME0 || (CANx->TSR&CAN_TSR_TME1) == CAN_TSR_TME1 || (CANx->TSR&CAN_TSR_TME2) == CAN_TSR_TME2) allow = true;
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_ReceiveErrorCount(const TinyCLR_Can_Provider* self, int32_t &errorCount) {
+    int32_t channel = self->Index;
+
+    CAN_TypeDef* CANx = ((channel == 0) ? CAN1 : CAN2);
+
+    errorCount = (uint8_t)((CANx->ESR & CAN_ESR_REC) >> 24);;
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_TransmitErrorCount(const TinyCLR_Can_Provider* self, int32_t &errorCount) {
+    int32_t channel = self->Index;
+
+    CAN_TypeDef* CANx = ((channel == 0) ? CAN1 : CAN2);
+
+    errorCount = (uint8_t)((CANx->ESR & CAN_ESR_REC) >> 16);;
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_GetSourceClock(const TinyCLR_Can_Provider* self, uint32_t &sourceClock) {
+    sourceClock = STM32F4_APB1_CLOCK_HZ;
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_SetReceiveBufferSize(const TinyCLR_Can_Provider* self, int32_t size) {
```

Comment:
If there is an internal transmit buffer, it should have a function to control its buffer as well.